### ISSUE TITLE
Torch Forecasting Models -  time spans

### ISSUE THREAD

**Issue by MichalRIcar** [unit8co/darts#1818]
Hello, 

I am exploring and experimenting with your truly amazing "the" arts and wonder about following (which is not described anywhere in docs or guides as far as I know):
1. I will use RNN as an example

2. We know from the design of RNN that batch size param is one of the more important ones as it is directly connected with gradient descent and general approch how we want to deal with a bias-variance trade off during training (fitting), e.g. https://machinelearningmastery.com/how-to-control-the-speed-and-stability-of-training-neural-networks-with-gradient-descent-batch-size/

3. While playing with your variation of RNN in darts lib I wonder how exactly do you combine training_length and input_chunk_length params with basic batch_size param?

4. I read your nice guide - https://unit8co.github.io/darts/userguide/torch_forecasting_models.html#top-level-look-at-training-and-predicting-with-chunks 

5. What I conclude from it is following: 
- Given provided batch_size → let's say 20, and 
- the training_length → assume of length 10 , together with 
- input_chunk_length, let's assume of 8
- For each batch pass the algo splits provided batch sample of 20 obs (defined by batch_size) to past and future whereas past data is where model fits and future data is where model computes objective function (e.g. RSS) in a fashion where training_length param takes different starting poits of sequence of 10 obs (from the sample size 20) and splits it to 8 past (defined by input_chunk_length  ) and 2 future. 
- Fit on 8, evaluate on 2 and do this 11 times as that is given by the limitation of provided sample (batch sample of length 20 in our example can provide 11 different samples of length 10 defined by the training_length).
- So that within each batch we have 11 fits and thus 11 RSS values. 
- If you crack the ts sequence then you can have much more than 11 fits inside 20 obs, but for the initial discussion I would assume the simpliest approach where you keep ts sequence withou shuffling.

6. So far so good as this approach is a standard cross-validation approach, however, where things get really unclear to me is a moment when I saw your examples of parametrization where you used batch_size = 8 and training_length = 24 and input_chunk_length =16 or so.

7. Another confusing part is in your docs → "Side note: Our *RNNModels accept a training_length parameter at model creation instead of output_chunk_length. Internally the output_chunk_length for these models is automatically set to 1." 

8. What is the reason of introducing new param training_length and "hide" output_chunk_length and set it to constat value of 1?  How this param setup works at the end? From this little side note it almost seems that training_length overrides batch_size and thus training_length is "your" batch_size with fixed value of  output_chunk_length  = 1 thus actually the whole "batching" or sampling of the total training sample is driven by training_length rather than batch_size. 

9. And to conclude further, the actual setup happens between training_length and input_chunk_length (whereas output_chunk_length = 1 in RNNModels) and thus batch_size is irrelevant for sampling in your param setup, thus it only serves as a signal to update NN weights when given batch_size value is reached. And reaching the batch_size value is from training_length sequential variantions as described above when algo computes RSS value and every time when new RSS values are equal to batch_size param is a moment when algo updates NN weights. 

Can you please clarify how algo works when input sample for given pass of training  (i.e. batch_size) is smaller than training_length and  even input_chunk_length ?

I played with different RNN setups with regards to these params and results are pretty different from each other and as a side note I am a ML guy developing own libraries and algos so I believe some docs upgrades are needed.

Thank you very much for the clarification.

**Comment by madtoinou** [comment]
Hi @MichalRIcar,

Thank you for this detailed question, I'll try to address it point by point:

1. `output_chunk_length=1` is enforced because by definition/design, RNN models predict values (tokens) one by one.
2. `training_length` is introduced because often, users are interested in predicting several values at once (a forecast with a horizon of 1 is rarely enough). Hence, this value should be used to create longer input and train the model in a context closer to what is happening at inference time.
3. `input_chunk_length` is used only at inference time, to slice the input appropriately (whereas `training_length` is used during training)
4. `batch_size` and `training_length` correspond to two different dimensions, `batch_size` correspond to the number of slices of the original `TimeSeries` used at once in a forward pass whereas `training_length` correspond to the length of the slices. Hence, they do not overwrite each other.

If you want to learn more about this, you can check at the source code of `RNNModel` but also `DualCovariatesInferenceDataset` [doc link](https://unit8co.github.io/darts/generated_api/darts.utils.data.inference_dataset.html?highlight=dualcovariate#darts.utils.data.inference_dataset.DualCovariatesInferenceDataset) and `GenericInferenceDataset [doc link](https://unit8co.github.io/darts/generated_api/darts.utils.data.inference_dataset.html?highlight=generic#darts.utils.data.inference_dataset.GenericInferenceDataset).

Let me know if some areas remain unclear
